test/unit/key_def.cc

```diff
@@ -457,16 +460,892 @@ test_tuple_validate_key_parts_raw(void)
+	plan(p);
+	header();
+
+	/* has_optional_parts is only valid if is_nullable. */
+	fail_unless(!has_optional_parts || is_nullable);
+
+	/* Type is number to prevent using precompiled comparators. */
+	const bool last_is_nullable = has_optional_parts;
+	struct key_def *key_def = test_key_def_new(
+		"[{%s%u%s%s}{%s%u%s%s}{%s%u%s%s%s%b}{%s%u%s%s%s%b}]",
+		"field", 0, "type", "number",
+		"field", 1, "type", "number",
+		"field", 2, "type", "number", "is_nullable", is_nullable,
+		"field", 3, "type", "number", "is_nullable", last_is_nullable);
+
+	/* Update has_optional_parts if the last parts can't be nil. */
+	size_t min_field_count = tuple_format_min_field_count(&key_def, 1,
+							      NULL, 0);
+	key_def_update_optionality(key_def, min_field_count);
+
+	fail_unless(key_def->is_nullable == is_nullable);
+	fail_unless(key_def->has_optional_parts == has_optional_parts);
+
+	const char *funcname = (const char *)xstrdup(tt_sprintf(
+		"tuple_compare_sequential<%s, %s>", is_nullable ? "true" :
+		"false", has_optional_parts ? "true" : "false"));
+
+	std::vector<struct tuple *> tuples_eq;
+	std::vector<struct tuple *> tuples_gt;
+
+	test_generate_common_cases(tuples_eq, tuples_gt, key_def);
+
+	test_check_cases(tuples_eq, tuples_gt, key_def, funcname,
+			 test_check_tuple_compare);
+
+	test_delete_cases(tuples_eq, tuples_gt);
+	key_def_delete(key_def);
+
+	footer();
+	check_plan();
+}
+
+static void
+test_tuple_compare_sequential_no_optional_parts_unique(bool is_nullable)
+{
+	plan(is_nullable ? 6 : 2);
+	header();
+
+	/* The primary key (PK). */
+	struct key_def *pk_def = test_key_def_new(
+		"[{%s%u%s%s}]",
+		"field", 3, "type", "number");
+
+	/* The secondary key (SK). */
+	struct key_def *key_def = test_key_def_new(
+		"[{%s%u%s%s}{%s%u%s%s%s%b}{%s%u%s%s}]",
+		"field", 0, "type", "number",
+		"field", 1, "type", "number", "is_nullable", is_nullable,
+		"field", 2, "type", "number");
+
+	struct key_def *cmp_def = key_def_merge(key_def, pk_def);
+	fail_unless(cmp_def->unique_part_count > key_def->part_count);
+
+	/*
+	 * It's assumed that PK and SK index different parts. So we
+	 * cover cmp_def->unique_part_count < cmp_def->part_count
+	 * branch of the sequential comparator (its last loop).
+	 */
+	cmp_def->unique_part_count = key_def->part_count;
+
+	/* Update has_optional_parts (the last parts can't be nil). */
+	struct key_def *keys[] = {pk_def, key_def};
+	size_t min_field_count = tuple_format_min_field_count(
+		keys, lengthof(keys), NULL, 0);
+	key_def_update_optionality(cmp_def, min_field_count);
+
+	fail_unless(cmp_def->is_nullable == is_nullable);
+	fail_unless(!cmp_def->has_optional_parts);
+
+	const char *funcname = (const char *)xstrdup(tt_sprintf(
+		"tuple_compare_sequential<%s, false, key_def is unique>",
+		is_nullable ? "true" : "false"));
+
+	/* Regular cases are tested in test_tuple_compare_sequential. */
+	std::vector<struct tuple *> tuples_eq;
+	std::vector<struct tuple *> tuples_gt;
```

Comment:
Why don't you generate common test cases here?

Reply:
The original problem was a key def difference, but it's not a problem after `test_generate_common_cases` refactoring.

Added the common cases testing.